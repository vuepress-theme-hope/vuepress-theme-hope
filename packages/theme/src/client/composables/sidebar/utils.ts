import { h } from "vue";
import { useRoute } from "vue-router";
import { useIconPrefix } from "../themeData";
import AutoLink from "../../components/AutoLink";
import { isActiveLink } from "../../utils";

import type { VNode } from "vue";
import type {
  ResolvedSidebarItem,
  ResolvedSidebarHeaderItem,
  ResolvedSidebarGroupItem,
} from "../../../shared";

export const renderIcon = (icon?: string): VNode | null =>
  icon
    ? h("i", {
        class: ["iconfont", `${useIconPrefix().value}${icon}`],
      })
    : null;

export const renderItem = (
  config: ResolvedSidebarItem,
  props: VNode["props"]
): VNode => {
  return config.link
    ? // if the item has link, render it as `<AutoLink>`
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      h(AutoLink, {
        ...props,
        config,
      })
    : // if the item only has text, render it as `<p>`
      h("p", props, [renderIcon(config.icon), config.text]);
};

export const renderGroupHeader = (
  { collapsable, icon, text }: ResolvedSidebarGroupItem,
  open: boolean
): (VNode | null)[] => [
  // icon
  renderIcon(icon),
  // title
  h("span", { class: "title" }, text),
  // arrow
  collapsable
    ? h("span", { class: ["arrow", open ? "down" : "right"] }, text)
    : null,
];

export const renderChildren = (
  children: ResolvedSidebarHeaderItem[]
): VNode | null => {
  const route = useRoute();
  if (!children) return null;

  return h(
    "ul",
    { class: "sidebar-sub-headers" },
    children.map((child) => {
      const active = isActiveLink(route, child.link);

      return h("li", { class: "sidebar-sub-header" }, [
        renderItem(child, {
          class: {
            "sidebar-link": true,
            heading: true,
            active,
          },
        }),
        renderChildren(child.children),
      ]);
    })
  );
};
