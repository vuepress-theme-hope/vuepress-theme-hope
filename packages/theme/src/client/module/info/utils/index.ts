export * from "./define";
export * from "./editlink";
