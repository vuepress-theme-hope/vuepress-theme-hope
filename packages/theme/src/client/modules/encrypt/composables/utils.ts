import { type ComputedRef, computed } from "vue";

import { useThemeData } from "@theme-hope/composables/index";

import { type EncryptConfig } from "../../../../shared/index.js";

export const useEncryptData = (): ComputedRef<EncryptConfig> => {
  const themeData = useThemeData();

  return computed(() => themeData.value.encrypt || {});
};
