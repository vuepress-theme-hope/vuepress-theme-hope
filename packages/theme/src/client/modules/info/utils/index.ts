export * from "./editLink.js";
export * from "./typings.js";
