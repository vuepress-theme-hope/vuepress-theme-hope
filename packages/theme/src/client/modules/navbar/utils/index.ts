import {
  type AutoLinkOptions,
  type NavGroup,
  type NavbarItem,
} from "../../../../shared/index.js";

// resolved
export type ResolvedThemeNavbarItem =
  | NavbarItem
  | NavGroup<AutoLinkOptions | NavGroup<AutoLinkOptions>>;
