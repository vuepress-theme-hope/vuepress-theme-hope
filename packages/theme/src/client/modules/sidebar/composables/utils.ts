import { ensureEndingSlash } from "@vuepress/shared";

export const resolvePrefix = (prefix = "", path = ""): string =>
  path.startsWith("/") ? path : `${ensureEndingSlash(prefix)}${path}`;
