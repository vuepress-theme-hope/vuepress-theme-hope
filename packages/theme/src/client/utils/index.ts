export * from "./getAncestorLinks.js";
export * from "./resolveLinkInfo.js";
