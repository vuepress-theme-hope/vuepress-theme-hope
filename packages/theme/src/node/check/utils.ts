export const PLUGIN_CHECKLIST = [
  ["@vuepress/plugin-active-header-links", "activeHeaderLinks"],
  ["@vuepress/plugin-blog", "blog"],
  ["@vuepress/plugin-catalog", "catalog"],
  ["@vuepress/plugin-comment", "comment"],
  ["@vuepress/plugin-copy-code", "copyCode"],
  ["@vuepress/plugin-copyright", "copyright"],
  ["@vuepress/plugin-docsearch", "docsearch"],
  ["@vuepress/plugin-external-link-icon", "externalLinkIcon"],
  ["@vuepress/plugin-feed", "feed"],
  ["@vuepress/plugin-git", "git"],
  ["@vuepress/plugin-links-check", "linksCheck"],
  ["@vuepress/plugin-notice", "notice"],
  ["@vuepress/plugin-nprogress", "nprogress"],
  ["@vuepress/plugin-photo-swipe", "photoSwipe"],
  ["@vuepress/plugin-prismjs", "prismjs"],
  ["@vuepress/plugin-pwa", "pwa"],
  ["@vuepress/plugin-reading-time", "readingTime"],
  ["@vuepress/plugin-redirect", "redirect"],
  ["@vuepress/plugin-rtl", "", 'Set "rtl: true" in the needed theme locales.'],
  ["@vuepress/plugin-search", "search"],
  ["@vuepress/plugin-seo", "seo"],
  ["@vuepress/plugin-sitemap", "sitemap"],
  ["@vuepress/plugin-theme-data", "", "This plugin is called internally."],
  ["@vuepress/plugin-watermark", "watermark"],
  ["vuepress-plugin-components", "components"],
  ["vuepress-plugin-md-enhance", "mdEnhance"],
  ["vuepress-plugin-search-pro", "searchPro"],
];
