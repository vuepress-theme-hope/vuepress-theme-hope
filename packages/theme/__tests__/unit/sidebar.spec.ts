import { createBaseApp } from "@vuepress/core";
import { path } from "@vuepress/utils";
import { describe, expect, it } from "vitest";

import { emptyTheme } from "./__fixtures__/theme/empty.js";
import { getStatus, getThemeData } from "../../src/node/config/index.js";
import { getSidebarData } from "../../src/node/prepare/index.js";
import { type ThemeOptions } from "../../src/shared/index.js";

describe("should generate sidebarData correctly", async () => {
  const app = createBaseApp({
    bundler: {} as any,
    source: path.resolve(__dirname, "./__fixtures__/src"),
    theme: emptyTheme,
  });

  await app.init();

  it("Should resolve structure in root", () => {
    const themeOptions: ThemeOptions = {
      sidebar: "structure",
    };

    const themeData = getThemeData(
      app,
      themeOptions,
      getStatus(app, themeOptions)
    );

    expect(getSidebarData(app, themeData)).toMatchSnapshot();
  });

  it("Should resolve structure in dir", () => {
    const themeOptions: ThemeOptions = {
      sidebar: {
        "/dir/": "structure",
        "/dir-negative-order/": "structure",
        "/dir-no-order/": "structure",
        "/link/": "structure",
        "/nested-dir/": "structure",
        "/nested-dir/nested-dir1/": "structure",
        "/options/": "structure",
      },
    };

    const themeData = getThemeData(
      app,
      themeOptions,
      getStatus(app, themeOptions)
    );

    expect(getSidebarData(app, themeData)).toMatchSnapshot();
  });

  it("Should resolve structure in group options", () => {
    const themeOptions: ThemeOptions = {
      sidebar: {
        "/nested-dir/": [
          "",
          "file1",
          {
            text: "nested-dir1",
            prefix: "nested-dir1/",
            children: "structure",
          },
        ],
      },
    };

    const themeData = getThemeData(
      app,
      themeOptions,
      getStatus(app, themeOptions)
    );

    expect(getSidebarData(app, themeData)).toMatchSnapshot();
  });
});
