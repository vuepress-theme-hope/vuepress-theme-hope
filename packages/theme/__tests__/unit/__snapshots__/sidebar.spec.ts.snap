// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should generate sidebarData correctly > Should resolve structure in dir 1`] = `
{
  "/dir-negative-order/": [
    "",
    "file2",
    "file4",
    "file3",
    "file1",
  ],
  "/dir-no-order/": [
    "",
    "file1",
    "file2",
    "file3",
    "file4",
  ],
  "/dir/": [
    "",
    "file2",
    "file4",
    "/permalink.html",
    "/2000/01/01/permalinkpattern/",
    "file1",
  ],
  "/link/": [
    "",
    "file2",
    "file4",
    "file1",
  ],
  "/nested-dir/": [
    "",
    "file2",
    {
      "children": [
        "",
        "file2",
        {
          "children": [
            "",
            "file2",
            "file4",
            "file1",
          ],
          "collapsible": true,
          "icon": "dir-icon",
          "prefix": "nested-dir2/",
          "text": "Dir2",
        },
        "file4",
        {
          "children": [
            "file2",
            "file4",
            "file1",
          ],
          "collapsible": true,
          "prefix": "nested-no-readme2/",
          "text": "Nested No Readme2",
        },
        "file1",
      ],
      "collapsible": true,
      "icon": "dir-icon",
      "prefix": "nested-dir1/",
      "text": "Dir1",
    },
    "file4",
    {
      "children": [
        "file2",
        "file4",
        {
          "children": [
            "file2",
            "file4",
            "file1",
          ],
          "collapsible": true,
          "prefix": "nested-no-readme2/",
          "text": "Nested No Readme2",
        },
        "file1",
      ],
      "collapsible": true,
      "prefix": "nested-no-readme1/",
      "text": "Nested No Readme1",
    },
    "file1",
  ],
  "/nested-dir/nested-dir1/": [
    "",
    "file2",
    {
      "children": [
        "",
        "file2",
        "file4",
        "file1",
      ],
      "collapsible": true,
      "icon": "dir-icon",
      "prefix": "nested-dir2/",
      "text": "Dir2",
    },
    "file4",
    {
      "children": [
        "file2",
        "file4",
        "file1",
      ],
      "collapsible": true,
      "prefix": "nested-no-readme2/",
      "text": "Nested No Readme2",
    },
    "file1",
  ],
  "/options/": [
    "",
    "file2",
    "file4",
    "file1",
  ],
}
`;

exports[`should generate sidebarData correctly > Should resolve structure in group options 1`] = `
{
  "/nested-dir/nested-dir1/": [
    "",
    "file2",
    {
      "children": [
        "",
        "file2",
        "file4",
        "file1",
      ],
      "collapsible": true,
      "icon": "dir-icon",
      "prefix": "nested-dir2/",
      "text": "Dir2",
    },
    "file4",
    {
      "children": [
        "file2",
        "file4",
        "file1",
      ],
      "collapsible": true,
      "prefix": "nested-no-readme2/",
      "text": "Nested No Readme2",
    },
    "file1",
  ],
}
`;

exports[`should generate sidebarData correctly > Should resolve structure in root 1`] = `
{
  "/": [
    "",
    {
      "children": [
        "",
        "file2",
        "file4",
        "/permalink.html",
        "/2000/01/01/permalinkpattern/",
        "file1",
      ],
      "collapsible": true,
      "icon": "dir-icon",
      "prefix": "dir/",
      "text": "Dir",
    },
    {
      "children": [
        "",
        "file2",
        "file4",
        "file3",
        "file1",
      ],
      "collapsible": true,
      "icon": "dir-icon",
      "prefix": "dir-negative-order/",
      "text": "Dir",
    },
    {
      "children": [
        "",
        "file1",
        "file2",
        "file3",
        "file4",
      ],
      "collapsible": true,
      "icon": "dir-icon",
      "prefix": "dir-no-order/",
      "text": "Dir",
    },
    {
      "children": [
        "",
        "file3",
        "file2",
        "file1",
        "file4",
      ],
      "collapsible": true,
      "icon": "dir-icon",
      "prefix": "dir-positive-order/",
      "text": "Dir",
    },
    {
      "children": [
        "",
        "file2",
        {
          "children": [
            "",
            "file2",
            {
              "children": [
                "",
                "file2",
                "file4",
                "file1",
              ],
              "collapsible": true,
              "icon": "dir-icon",
              "prefix": "nested-dir2/",
              "text": "Dir2",
            },
            "file4",
            {
              "children": [
                "file2",
                "file4",
                "file1",
              ],
              "collapsible": true,
              "prefix": "nested-no-readme2/",
              "text": "Nested No Readme2",
            },
            "file1",
          ],
          "collapsible": true,
          "icon": "dir-icon",
          "prefix": "nested-dir1/",
          "text": "Dir1",
        },
        "file4",
        {
          "children": [
            "file2",
            "file4",
            {
              "children": [
                "file2",
                "file4",
                "file1",
              ],
              "collapsible": true,
              "prefix": "nested-no-readme2/",
              "text": "Nested No Readme2",
            },
            "file1",
          ],
          "collapsible": true,
          "prefix": "nested-no-readme1/",
          "text": "Nested No Readme1",
        },
        "file1",
      ],
      "collapsible": true,
      "icon": "dir-icon",
      "prefix": "nested-dir/",
      "text": "Dir",
    },
    {
      "children": [
        "file2",
        "file4",
        "file1",
      ],
      "collapsible": true,
      "icon": "my-icon",
      "link": "link/",
      "prefix": "link/",
      "text": "Dir with link",
    },
    {
      "children": [
        "",
        "file2",
        "file4",
        "file1",
      ],
      "collapsible": true,
      "icon": "my-icon",
      "prefix": "options/",
      "text": "Dir with options",
    },
    {
      "children": [
        "component",
        "cut",
        "demo",
        "example",
        "link",
      ],
      "collapsible": true,
      "prefix": "excerpt/",
      "text": "Excerpt",
    },
  ],
}
`;
